---
title: "Basis of logistic regression"
author: "Oscar Gonzalez-Recio"
date: "May 2021"
output:
  html_document: 
    code_folding: hide
    collapsed: yes
    fig_caption: yes
    fig_width: 6
    highlight: tango
    number_sections: yes
    theme: united
    toc: yes
    toc_float: yes
  pdf_document: # options pour sorties pdf
    toc: yes
    toc_depth: '3'
  word_document: default
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

require(prettydoc)
require(ggplot2)

library(prettydoc)

# Generate data 
#GENERATOR MODEL
n=1000                            #ndata
x<-floor(runif(n,0,3))            #genotypes
beta<-rnorm(1,0,.8)              #allele subst effects
beta=2
mu=20                             #population mean
y<-mu+beta*x+rnorm(n,0,sd=5)      #generate data
y_cat<-ifelse(y<mean(y),0,1)      #categorize variable
data<-data.frame(y_cat,x)
```



# 1.WHAT IS LOGISTIC REGRESSION

Logistic regresion is a statistical model applied on binary outcome variables. It assumes that the logit transformation of the outcome variable has a linear relationship with the predictor variables.  This makes the interpretation of the regression coefficients somewhat tricky.  This document briefly explains how to interpret the logistic regression coeffcients and the odds ratio.

# 2. Logistic regression in GWAS

Lets assume a phenotype that is categorized in malignant (y=1) and no-malignant (y=0). The covariate are the SNP genotypes (AA, Aa and aa) labeled as 00, 01 and 11, respectively.

A scatter plot of the data is not very informative.

```{r,echo=F,warning=F,cache.comments=F}
#Calculate regression coeficients
plot(x,y_cat,pch=4,col=2,cex=2,xlab="Genotype",ylab="Malignant?")

```

We can make the plot look a bit nicer:


```{r,echo=F,warning=F,cache.comments=F}
ggplot(data=data, aes(x=x, y=y_cat)) + geom_point(position=position_jitter(h=0.1, w=0.1),shape = 21, alpha = 0.5, size = 3)

```

The incidence table is:
```{r,echo=F,warning=F,cache.comments=F}
#Calculate regression coeficients
library(dplyr)
data %>%
    count(x,y_cat, name="n")
tb<-tapply(y_cat,x,table)
perc00<-(tb[[1]][[2]]/(tb[[1]][[2]]+tb[[1]][[1]]))
perc11<-(tb[[3]][[2]]/(tb[[3]][[2]]+tb[[3]][[1]]))
```

We observe $$perc_0=`r round(perc00*100,1)`% $$ of malignant phenotypes in the subpopulation of individuals with the 00 genotype, and $$perc_2=`r round(perc11*100,1)`%$$ for individuals with the ALT homozygous genotype (11 or 2).


## Calculating the coefficients in logistic regression manually



Assuming that the 0 (0/0) genotype is the reference group, the ratio for this genotype between malignants and no-malignants is `r tb[[1]][[2]]` / `r tb[[1]][[1]]`=`r tb[[1]][[2]] /tb[[1]][[1]]`.

The Intercept in the log model is the logarithim of the above ratio: log(`r tb[[1]][[2]] /tb[[1]][[1]]`)= `r log(tb[[1]][[2]]/tb[[1]][[1]])`.

Then we can estimate the coefficient for the heterozygous (01) as:
$$
\log\frac{\frac{`r tb[[2]][[2]]`}{`r tb[[2]][[1]]`}}{\frac{`r tb[[1]][[2]]`}{`r tb[[1]][[1]]`}} =`r round(log( (tb[[2]][[2]]/tb[[2]][[1]]) / (tb[[1]][[2]]/tb[[1]][[1]]) ),3)`
$$

and for alternative homozygous (11) as .

$$
\log\frac{\frac{`r tb[[3]][[2]]`}{`r tb[[3]][[1]]`}}{\frac{`r tb[[1]][[2]]`}{`r tb[[1]][[1]]`}} =`r round(log( (tb[[3]][[2]]/tb[[3]][[1]]) / (tb[[1]][[2]]/tb[[1]][[1]]) ),3)`
$$



The Odds-ratio are calculated taking the exponential of these coefficients. For instance, the Odds-ratio for having a malignant phenotype between the 01 genotype and the 00 genotype is 
```{r,echo=F,warning=F,cache.comments=F}
or01<-(tb[[2]][[2]]/tb[[2]][[1]]) / (tb[[1]][[2]]/tb[[1]][[1]])

```


$$OR_1=\exp(`r log(or01)`) =`r round(or01,4)`$$

The Odds-ratio between 00 and 11 for malignant is 
```{r,echo=F,warning=F,cache.comments=F}
 or11<-(tb[[3]][[2]]/tb[[3]][[1]]) / (tb[[1]][[2]]/tb[[1]][[1]])
```

$$OR_2=\exp(`r log(or11)`)=`r round(or11,4)`$$
 This should be interpreted in terms of the reference group (00) which has probability of a malignant phenotype of being malignant of
 (`r tb[[1]][[2]]`)/(`r tb[[1]][[1]]+tb[[1]][[2]]`)=`r tb[[1]][[2]]/(tb[[1]][[2]]+tb[[1]][[1]])`.
 
 Then, with a little bit of algebra, we can calculate the expected proportion of indviduals showing malignant phenotype in the 11 genotype group as 
 $$
 perc_2=\frac{perc_0*or11}{1+perc_0(or11-1)}=`r perc00*or11/(1+(or11-1)*perc00)`
 $$
which in this case is equal to perc_2 (`r round(100*perc11,1)`)%.
 


## The automatic model

We can use the glm function in R, using *family= binomial* option, and check the results. We must point out that more logistic regression models with more variables need of more complex solving algorithms

```{r,echo=F,warning=F,cache.comments=F}
#Calculate regression coeficients
log_reg<-glm(y_cat~as.factor(x),family = binomial)
summary(log_reg)

```

