---
title: "GWAS - continuous trait - population structure"
author: "Filippo Biscarini"
date: "October 28, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("knitr")
library("qqman")
library("dplyr")
library("ggplot2")
library("snpStats")
library("ggfortify")
library("doParallel")

source("../software/GWASfunction.R")

load("genData_continuous.RData")
```

## GWAS

### Set up

```{r, echo=TRUE}
# Choose trait for association analysis, use colnames(genData$LIP) for listing
# NOTE: Ignore the first column of genData$LIP (gender)
target <- "PH"
phenodata <- data.frame("id" = rownames(genData$PHENOTYPES),
                        "phenotype" = genData$PHENOTYPES[,target], stringsAsFactors = F)
trait_type = "gaussian"
```

## Population structure

We can first have a look at the data to see if there is any obvious population structure. Plotting the first two principal components is a quick and easy way to do this.

```{r, echo=TRUE}
write.SnpMatrix(x = genData$SNP, file = "temp.txt", as.alleles = FALSE, quote = FALSE)
x <- read.table("temp.txt", header = FALSE, skip = 1)

autoplot(prcomp(x[,-1]), data = phenodata, colour = 'phenotype')
phenodata$group <- genData$PHENOTYPES$population
autoplot(prcomp(x[,-1]), data = phenodata, colour = 'group')
```

## GWAS with family structure as fixed effects

We have now added a column to the data frame `phenodata`: this way, we are telling the function GWAA to run a model that includes "population" as a systematic effect (classes):

$$
y_{ijk} = \mu + \text{population}_k + \text{SNP}_j \cdot x_{ji} + e_{ijk}
$$

This is because GWAA will call a GLM (generalised linear model) with the following *R* syntax: `phenotype~ . - id`

```{r, echo=TRUE}
kable(head(phenodata))
```

```{r, echo=TRUE}
start <- Sys.time()
GWAA(genodata = genData$SNP, phenodata = phenodata, family = trait_type, filename = paste(target, ".txt", sep = ""))
Sys.time() - start # benchmark
```

### Results

```{r, echo=TRUE}
GWASout <- read.table(paste(target, ".txt", sep = ""), header = T, colClasses = c("character", rep("numeric",4)), dec=".")
GWASout$type <- rep("typed", nrow(GWASout))
GWASout$Neg_logP <- -log10(GWASout$p.value)
GWASout <- merge(GWASout, genData$MAP[,c("SNP", "chr", "position")])
GWASout <- GWASout[order(GWASout$Neg_logP, decreasing = T),]

gwasResults <- GWASout[,c("SNP","chr","position","p.value","t.value")]
names(gwasResults) <- c("SNP","CHR","BP","P","zscore")
manhattan(gwasResults, suggestiveline = FALSE)
qq(gwasResults$P)

## calculates lambda
z=qnorm(gwasResults$P/2)
lambda = round(median(z^2,na.rm=T)/qchisq(0.5,df=1),3)
print(lambda)
```

Lambda was reduced to **`r lambda`**, which points to an improved fit of the model (lower chance of spurious results).


##  Conduct GWAS with PCs to account for family structure in the model

Another way to account for population structure in a GWAS analysis is:

1. calculate principal components
2. add the first 3 principal components as systematic effects to the model
3. run the model

```{r, echo=TRUE}
pc <- prcomp(x[,-1], rank. = 3)
phenodata <- cbind.data.frame(phenodata,pc$x)
phenodata$group <- NULL
```

We see that now the `phenodata` dataframe features three new columns: PC1, PC2, PC3

```{r, echo=TRUE}
kable(head(phenodata))
```

```{r, echo=TRUE}
start <- Sys.time()
GWAA(genodata = genData$SNP, phenodata = phenodata, family = trait_type, filename = paste(target, ".txt", sep = ""))
Sys.time() - start # benchmark
```

### Results

```{r, echo=TRUE}
GWASout <- read.table(paste(target, ".txt", sep = ""), header = T, colClasses = c("character", rep("numeric",4)), dec=".")
GWASout$type <- rep("typed", nrow(GWASout))
GWASout$Neg_logP <- -log10(GWASout$p.value)
GWASout <- merge(GWASout, genData$MAP[,c("SNP", "chr", "position")])
GWASout <- GWASout[order(GWASout$Neg_logP, decreasing = T),]

gwasResults <- GWASout[,c("SNP","chr","position","p.value","t.value")]
names(gwasResults) <- c("SNP","CHR","BP","P","zscore")
manhattan(gwasResults, suggestiveline = FALSE)
qq(gwasResults$P)

## calculates lambda
z=qnorm(gwasResults$P/2)
lambda = round(median(z^2,na.rm=T)/qchisq(0.5,df=1),3)
print(lambda)
```

Lambda is further reduced to **`r lambda`**, which points to an improved fit of the model (lower chance of spurious results).
