---
title: "GWAS - binary trait"
author: "Filippo Biscarini"
date: "October 10, 2019"
output: html_document
bibliography: gwas.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("knitr")
library("qqman")
library("dplyr")
library("ggplot2")
library("snpStats")
library("doParallel")

source("../software/GWASfunction.R")
```

## GWAS for a binary trait

From Wolf et al. @wolf2015genome, data on cleft lip (affected / unaffected) in dogs (*Nova Scotia Duck Tolling Retriever - NSDTR*) have been obtained.

### Preparing the data

#### Genotypes

```{r data, echo=TRUE}
## Prepare data
pathD <- paste("../3.imputation/dogs_imputed", c(".bed", ".bim", ".fam"), sep = "")
genotypes <- read.plink(pathD[1], pathD[2], pathD[3])

map <- genotypes$map
colnames(map) <- c("chr", "SNP", "gen.dist", "position", "A1", "A2")
```

N. **`r nrow(map)` SNPs** from **`r nrow(genotypes$genotypes)` dog samples**. SNPs are distributed over **`r length(unique(map$chr))` chromosomes** (chr: `r unique(map$chr)`)

#### Phenotypes

```{r, echo=TRUE}
## phenotypes
phenotypes = read.table("../data/dogs_phenotypes.txt", header = TRUE)
names(phenotypes) <- c("id", "population", "condition")
```

Condition status (presence/absence of cleft lip) was recorded on `r nrow(phenotypes)` dog samples. Genotype records and phenotype records need in the first place to be aligned:

```{r, echo=TRUE}
phenotypes <- phenotypes[phenotypes$id %in% row.names(genotypes$genotypes),]
row.names(phenotypes) <- phenotypes$id
phenotypes$condition <- as.factor(phenotypes$condition)
phenotypes$id <- NULL
```

Data (both phenotypes and genotypes) are then saved to a list and written out for next steps:

```{r, echo=TRUE}
genData = list(SNP=genotypes$genotypes, MAP=map, PHENOTYPES=phenotypes)
save(genData, file = "genData_binary.RData")

### analysis
load("genData_binary.RData")
```

One breed, two phenotypic categories:

- **1** = cleft lip (malformation)
- **0** = healthy (control)

Below the distribution of phenotypes (PH) after alignment to genotypes:

```{r, echo=TRUE}
D <- genData$PHENOTYPES %>%
  group_by(condition) %>%
  summarise(N=n())

kable(D)
```

```{r, echo=TRUE}
p <- ggplot(genData$PHENOTYPES, aes(x=as.factor(condition), fill=as.factor(condition)))
p <- p + geom_bar() + xlab("condition (cleft lip")
p
```

## Data filtering

```{r, echo=TRUE}
# Use SNP call rate of 100%, MAF of 0.1 (very stringent)
maf <- 0.05
callRate <- 1
SNPstats <- col.summary(genData$SNP)

maf_call <- with(SNPstats, MAF > maf & Call.rate == callRate)
genData$SNP <- genData$SNP[,maf_call]
genData$MAP <- genData$MAP[maf_call,]
SNPstats <- SNPstats[maf_call,]
```

## GWAS

### Set up

```{r, echo=TRUE}
# Choose trait for association analysis, use colnames(genData$LIP) for listing
# NOTE: Ignore the first column of genData$LIP (gender)
target <- "condition"
phenodata <- data.frame("id" = rownames(genData$PHENOTYPES),
                        "phenotype" = genData$PHENOTYPES[,target], stringsAsFactors = F)
trait_type = "binomial"
```

### Execution

```{r, echo=TRUE}
# Conduct GWAS (will take a while)
start <- Sys.time()
GWAA(genodata = genData$SNP, phenodata = phenodata, family = trait_type, filename = paste(target, ".txt", sep = ""))
Sys.time() - start # benchmark
```

### Results

```{r, echo=TRUE}
GWASout <- read.table(paste(target, ".txt", sep = ""), header = T, colClasses = c("character", rep("numeric",4)), dec=".")
GWASout$Neg_logP <- -log10(GWASout$p.value)
GWASout <- merge(GWASout, genData$MAP[,c("SNP", "chr", "position")])
GWASout <- GWASout[order(GWASout$Neg_logP, decreasing = T),]

gwasResults <- GWASout[,c("SNP","chr","position","p.value","t.value")]
names(gwasResults) <- c("SNP","CHR","BP","P","zscore")

kable(head(gwasResults))
```

#### Manhattan plot

```{r, echo=TRUE}
manhattan(gwasResults)
```

#### QQ plot

```{r, echo=TRUE}
qq(gwasResults$P)
```

#### Inflation factor $\lambda$

```{r, echo=TRUE}
## calculates lambda
z=qnorm(gwasResults$P/2)
lambda = round(median(z^2,na.rm=T)/qchisq(0.5,df=1),3)
print(lambda)
```



## References
