---
title: "GWAS - binary trait - kinship"
author: "Filippo Biscarini"
date: "October 28, 2019"
output: html_document
---

```{r setup, include=FALSE}
### KINSHIP matrix
source("../software/gwas.r")
source("../software/emma.r")

library("qqman")
library("gMatrix") ## own-made package
```

### Prepare the data

```{r, echo=TRUE}
snpMatrix <- read.table("../3.imputation/dogs_imputed.raw", header = TRUE)
SNP_INFO <- read.table("../3.imputation/dogs_imputed.map")
names(SNP_INFO) <- c("Chr","SNP","cM","Pos")

phenotypes <- read.table("../data/dogs_phenotypes.txt", header = TRUE)
phenotypes <- phenotypes[phenotypes$id %in% snpMatrix$IID,]
```

```{r, label='genotype_matrix'}
X <- as.matrix(snpMatrix[,-c(1:6)])
colnames(X) <- gsub("\\_[A-Z]{1}$","",colnames(X))
rownames(X) <- snpMatrix$IID
```

### Kinship matrix

A more sophisticated way to account for finer population structure is to use the **kinship matrix** as covariance structure in a **linear mixed model**.
The kinship matrix (a.k.a. additive relationship matrix) can be calculated in several different ways: here we use the approach by Van Raden 2008 (add ref)

```{r heatmap}
K <- gVanRaden(X)
heatmap(K,col=rev(heat.colors(75)))
```

## GWAS with the kinship matrix to account for population structure

```{r}
Y <- as.matrix(phenotypes$phenotype)
rownames(Y) <- phenotypes$id

res <- amm_gwas(Y = Y, X = X, K = K, m = 1, use.SNP_INFO = TRUE)
```

### Results

```{r}
gwasResults <- res[,c("SNP","Chr","Pos","Pval")]
names(gwasResults) <- c("SNP","CHR","BP","P")
manhattan(gwasResults, suggestiveline = FALSE, col = c("red","blue"))
qq(gwasResults$P)
```

```{r}
z=qnorm(gwasResults$P/2)
lambda = round(median(z^2,na.rm=T)/qchisq(0.5,df=1),3)
lambda
```

Inflation is reduce to **`r lambda`**
