---
title: "GWAS basic models"
author: "Filippo Biscarini"
date: "October 26, 2019"
output:
  ioslides_presentation:
      css: styles.css
  slidy_presentation: default
---

```{r setup, include=FALSE}
library("knitr")
library("ggplot2")

knitr::opts_chunk$set(echo = FALSE)
knitr::knit_hooks$set(
   error = function(x, options) {
     paste('\n\n<div class="alert alert-danger">',
           gsub('##', '\n', gsub('^##\ Error', '**Error**', x)),
           '</div>', sep = '\n')
   },
   warning = function(x, options) {
     paste('\n\n<div class="alert alert-warning">',
           gsub('##', '\n', gsub('^##\ Warning:', '**Warning**', x)),
           '</div>', sep = '\n')
   },
   message = function(x, options) {
     paste('\n\n<div class="alert alert-info">',
           gsub('##', '\n', x),
           '</div>', sep = '\n')
   }
)

```

## GWAS: quantitative traits

For quantitative traits (continuous target variables), the basic model is a simple linear regression model of the following form:

$$
y_{ij}=\mu+SNP_j+e_{ij}
$$

where $y_{ij}$ is the phenotypic observation for sample *i* with SNP genotype *j*, $SNP_j$ is the effect of SNP genotype *j* (e.g. AA, AB, BB), and $e_{ij}$ are the residuals of the model.

## Loading the data

```{r data, echo=TRUE}
phenotypes <- read.table(
  "../data/rice_phenotypes.txt", 
  header = TRUE)

genotypes <- read.table(
  "../3.imputation/rice_imputed.raw", 
  header = TRUE)

genotypes <- genotypes[,-c(1,3,4,5,6)]
# genotypes[,c(2,7)]

df <- merge(phenotypes, genotypes[,c(1,2)], by.x="id", by.y="IID")
names(df) <- c("id","population","PH","snp")
```

## head(df)

```{r}
kable(head(df))
```

## Modelling number of copies of the minor allele

First, we can model SNP genotypes as the number of copies of the minor allele: 0, 1, 2 (numeric variable)

```{r model1, echo=TRUE}
fit <- lm(PH~snp, data = df)
```

## Results - summary(fit)

```{r, echo=FALSE}
g <- summary(fit)
print(g)
```

## ANOVA table

```{r, echo=TRUE}
kable(anova(fit))
```

## Modelling genotypes

Genotypes can be coded as AA, AB and BB (factor variables), instead. The p-value is not affected, but the estimated SNP effect is.

```{r model2, echo=TRUE}
df$snp <- as.factor(df$snp)
fit <- lm(PH~snp, data = df)
```

## Results - summary(fit)

```{r, echo=FALSE}
g <- summary(fit)
print(g)
```

## ANOVA table

```{r, echo=FALSE}
kable(anova(fit))
```

## Adding population structure

- the rice dataset comprises records from **`r length(unique(df$population))` rice populations** (*`r unique(df$population)`*)
- we can (and should) model **population structure**

```{r, echo=TRUE}
fit <- lm(PH~population+snp, data = df)
```

## Results - summary(fit)

```{r, echo=FALSE}
g <- summary(fit)
print(g)
```

## ANOVA table

```{r}
kable(anova(fit))
```

# GWAS: binary traits

## GWAS: the binary trait model

For binary traits (e.g. case/control), the basic model is a simple logistic regression model of the following form:

$$
\left\{ 
\begin{array}{lr}
P(y_i=1|x_i)=p(x_i) \\
\text{logit}(p(x_i))=\text{log} \left( \frac{p(x_i)}{1-p(x_i)} \right)=\mu+\sum_{j=1}^m z_{ij} SNP_j
\end{array}
\right.
$$

where $P(y_i=1|x_i)$ is the probability for sample *i* to belong to group "1" given its SNP genotypes *x_i*.

## Loading the data

```{r data_sheep, echo=TRUE}
phenotypes <- read.table(
  "../data/dogs_phenotypes.txt", 
  header = TRUE)
genotypes <- read.table(
  "../3.imputation/dogs_imputed.raw", 
  header = TRUE)
genotypes <- genotypes[,-c(1,3,4,5,6)]

df <- merge(phenotypes, genotypes[,c(1,25)], by.x="id", by.y="IID")
names(df) <- c("id","family","disease","snp")
```

## head(df)

```{r,echo=FALSE}
kable(head(df))
```

## Fit a generalised linear model

We now need to fit a generalised linear model instead of a simple linear regression model: since we are analysing binary data, a common choice is to use the **binomial distribution** and the **logit()** link function: 

```{r, echo=TRUE}
fit <- glm(disease~snp, data = df, 
           family=binomial(link="logit"))
```

## Results - summary(fit) {.smaller}

```{r, echo=FALSE}
g <- summary(fit)
print(g)
```

## full vs reduced model

```{r, echo=TRUE}
full_model <- glm(disease~snp, data = df, 
                  family="binomial")
reduced_model <- glm(disease~1, data = df, 
                     family="binomial")
anova(reduced_model, full_model, test = "LRT")
```

## likelihood ratio test

```{r, echo=TRUE}
l1 <- logLik(full_model)
l0 <- logLik(reduced_model)
degrees_of_freedom <- length(coef(full_model)) - length(coef(reduced_model)) 
```

## chi-square statistic

```{r, echo=TRUE}
teststat<--2*(as.numeric(l0)-as.numeric(l1))
pchisq(teststat,df=degrees_of_freedom,lower.tail=FALSE)
```

## Add population structure

```{r, echo=TRUE, error=TRUE}
fit <- glm(disease~snp+family, data = df, 
           family=binomial(link="logit")) 
```

<div class="emphasized">
!! ERROR !! why?
</div>

## PC's

```{r, echo=TRUE}
pc <- prcomp(genotypes[,-1], rank. = 3)
df <- cbind.data.frame(df,pc$x)
```

## Eigenvalues

```{r, echo=TRUE}
eig <- (pc$sdev)^2
variance <- eig*100/sum(eig)
```

## Barplot of eigenvalues

```{r, echo=TRUE}
barplot(variance, col="red", xlab = "Principal Components", 
        ylab = "% variance")
sum(variance[c(1:3)])
```

## Use principal components to model population structure

```{r, echo=TRUE}
fit <- glm(
  disease~snp+PC1+PC2+PC3, 
  data = df, 
  family=binomial(link="logit")
)
```

## Results - summary(fit) {.smaller}

```{r, echo=FALSE}
g <- summary(fit)
print(g)
```

## full vs reduced model

```{r, echo=TRUE}
full_model <- glm(disease~snp+PC1+PC2+PC3, data = df, family="binomial")
reduced_model <- glm(disease~PC1+PC2+PC3, data = df, family="binomial")
anova(reduced_model, full_model, test = "LRT")
```

## likelihood ratio test

```{r, echo=TRUE}
l1 <- logLik(full_model)
l0 <- logLik(reduced_model)
degrees_of_freedom <- length(coef(full_model)) - 
  length(coef(reduced_model)) 
```

## chi-square statistic

```{r, echo=TRUE}
teststat<--2*(as.numeric(l0)-as.numeric(l1))
pchisq(teststat,df=degrees_of_freedom,lower.tail=FALSE)
```


